### ISSUE TITLE
Haswell reported as Nehalem...

### ISSUE THREAD

**Issue by migueldiascosta** [archspec/archspec#40]
...because the `aes` feature doesn't seem to be present in this particular host (perhaps because of export restrictions?)

Could archspec (optionally?) ignore the absence of `aes`?

**Comment by alalazo** [comment]
Sorry for the late reply. I don't think we can do that since GCC explicitly mentions that when `-march=haswell` is set the binary may have AES instructions.

**Comment by boegel** [comment]
@alalazo The "may" here already indicates that it's not guaranteed?

If archspec is currently misidentifying Haswell (AVX2) as Nehalem (which doesn't even have AVX!) solely based on `aes`, then that's a problem imho...

**Comment by alalazo** [comment]
May means that your code may not need that instruction, but GCC is allowed to emit it if it thinks it's needed. So my understanding is that there's potential for incompatibility here. We need a finer control if we want to be on a `haswell` without `aes`.

**Comment by alalazo** [comment]
For reference:
```
westmere
    Intel Westmere CPU with 64-bit extensions, MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, POPCNT, AES and PCLMUL instruction set
    support.
```

**Comment by tgamblin** [comment]
We don't want `archspec` to tell tools to build binaries that are incompatible with the host.  If the host says it doesn't support a particular instruction, and we circumvent that somehow and provide flags that result in the compiler emitting that instruction, things will break.

This is probably a good place to start considering x86_64 levels archspec/archspec-json#31.  I suspect that a lot of users, like @migueldiascosta, do not care about AES, but just want the vector instructions. So what you're rather have happen here is for archspec to identify the machine as `x86_64_v3` because it has `avx2`, instead of downgrading to Nehalem.

I think our detection algorithm currently starts with `x86_64` and goes up until it finds the `uarch` with the maximal subset of the host arch's features.

Why don't we change it to something like:

1. Start with `x86_64v4`, test newer architectures until we find a maximal subset, return it if found.
2. Start with `x86_64v3`, test newer architectures until we find a maximal subset, return it if found.
3. Start with `x86_64v2`, test newer architectures until we find a maximal subset, return it if found.
4. etc.

Basically start with the newest "generic" architecture, then try to be more specific from there, but keep the generics as the baselines.

This would mean that when features we don't really care about are disabled by VMs, we still get the vector instructions we care about if they're available (that is what most people are after).  I think we could do the same thing with ARM architecture levels -- we'd end up with, e.g., `armv8` instead of always reverting to plain old `aarch64`.

Thoughts?

**Comment by alalazo** [comment]
That seems a good way to go now that we have `x86_64_vX`. Minor shuffling of the algorithm proposed above:
1. Test only `x86_64_vX` generic archs, record the best one that match
2. Test "real" uarch like we do now only if `uarch > best_generic_match`
3. Return the best uarch found at point 2 or the generic at point 1 if 2 has no match

**Comment by tgamblin** [comment]
Sounds good to me!  We can hash the details out on a PR but seems like we're proposing the same thing.

**Comment by alalazo** [comment]
I'll take a stab at that tomorrow.

**Comment by migueldiascosta** [comment]
minor comment, this was on a physical host, not a VM, so there really are Haswell's out there (or here, at least :) ) without `aes`

but yes, with the possibility of VMs disabling features the levels become even more important

**Comment by tgamblin** [comment]
Yep -- AES is customizable at the BIOS level: https://techlibrary.hpe.com/docs/iss/proliant_uefi/UEFI_TM_030617/s_enable_AES-NI.html.

**Comment by migueldiascosta** [comment]
ah, so maybe that Haswell is not really "crippled" like I assumed, ok :)

you're right, bios customizable features are another reason to use the architecture levels

**Comment by alalazo** [comment]
@migueldiascosta Do you still have access to that machine? If so can you test drive #53 there?

**Comment by migueldiascosta** [comment]
@alalazo yep!

```bash
$ git checkout master
Switched to branch 'master'
$ python -c "from archspec import cpu; print(cpu.host())"
nehalem
$ python -c "from archspec import cpu; print('avx2' in cpu.host())"
False
$ git checkout features/detection_algorithm
M	archspec/json
Switched to branch 'features/detection_algorithm'
$ python -c "from archspec import cpu; print(cpu.host())"
x86_64_v3
$ python -c "from archspec import cpu; print('avx2' in cpu.host())"
True
```
